src/Nest/Domain/Observers/ReindexObservable.cs

```diff
@@ -1,115 +1,125 @@
-				this._reindexDescriptor._CreateIndexSelector ?? ((ci) => ci);
-
-			var createIndexResponse = this.CurrentClient.CreateIndex(
-				toIndex, (c) => settings(c.InitializeUsing(indexSettings.IndexSettings)));
-			if (!createIndexResponse.IsValid)
-				throw new ReindexException(createIndexResponse.ConnectionStatus);
-
-			var page = 0;
-			var searchResult = this.CurrentClient.Search<T>(
-				s => s
-					.Index(fromIndex)
-					.AllTypes()
-					.From(0)
-					.Size(size)
-					.Query(this._reindexDescriptor._QuerySelector ?? (q=>q.MatchAll()))
-					.SearchType(SearchType.Scan)
-					.Scroll(scroll)
-				);
-			if (searchResult.Total <= 0)
-				throw new ReindexException(searchResult.ConnectionStatus, "index " + fromIndex + " has no documents!");
-			IBulkResponse indexResult = null;
-			do
-			{
-				var result = searchResult;
-				searchResult = this.CurrentClient.Scroll<T>(s => s
-					.Scroll(scroll)
-					.ScrollId(result.ScrollId)
-				);
-				if (searchResult.Documents.HasAny())
-					indexResult = this.IndexSearchResults(searchResult, observer, toIndex, page);
-				page++;
-			} while (searchResult.IsValid && indexResult != null && indexResult.IsValid && searchResult.Documents.HasAny());
-
-
-			observer.OnCompleted();
-		}
-
-		public IBulkResponse IndexSearchResults(ISearchResponse<T> searchResult,IObserver<IReindexResponse<T>> observer, string toIndex, int page)
-		{
-			if (!searchResult.IsValid)
-				throw new ReindexException(searchResult.ConnectionStatus, "reindex failed on scroll #" + page);
-
-			var bb = new BulkDescriptor();
-			foreach (var d in searchResult.Hits)
-			{
-				IHit<T> d1 = d;
-				bb.Index<T>(bi => bi.Document(d1.Source).Type(d1.Type).Index(toIndex).Id(d.Id));
-			}
-
-			var indexResult = this.CurrentClient.Bulk(b=>bb);
-			if (!indexResult.IsValid)
-				throw new ReindexException(indexResult.ConnectionStatus, "reindex failed when indexing page " + page);
-
-			observer.OnNext(new ReindexResponse<T>()
-			{
-				BulkResponse = indexResult,
-				SearchResponse = searchResult,
-				Scroll = page
-			});
-			return indexResult;
-		}
-
-
-		public void Dispose()
-		{
-
-		}
-	}
+			var size = this._reindexDescriptor._Size ?? 100;
+
+			fromIndex.ThrowIfNullOrEmpty("fromIndex");
+			toIndex.ThrowIfNullOrEmpty("toIndex");
+
+			var indexSettings = this.CurrentClient.GetIndexSettings(i => i.Index(this._reindexDescriptor._FromIndexName));
+			Func<CreateIndexDescriptor, CreateIndexDescriptor> settings =
+				this._reindexDescriptor._CreateIndexSelector ?? ((ci) => ci);
+
+			var createIndexResponse = this.CurrentClient.CreateIndex(
+				toIndex, (c) => settings(c.InitializeUsing(indexSettings.IndexSettings)));
+			if (!createIndexResponse.IsValid)
+				throw new ReindexException(createIndexResponse.ConnectionStatus);
+
+			var page = 0;
+			Func<SearchDescriptor<T>, SearchDescriptor<T>> searchDescriptor = s => s.Index(fromIndex);
+
+			if (typeof(T).Name.Equals(typeof(object).Name))

```

Comment:
Here: https://github.com/elastic/elasticsearch-net/blob/1433-support-generic-reindex/src/Nest/ElasticClient-Reindex.cs#L19 we added a non generic overload to the client. However we can not distinguish here between someone calling the generic or non generic overload. If someone now calls `client.Reindex<object>()` it will reindex on everything where as only `client.Reindex()` should.